Add support for incremental compilation

jackkoenig opened the pull request:
This switches from using Bloop to using mill as the underling compilation tool. It's not great, but it works and has incremental compile.

I've tried this on some complicated designs and it seems to work but need to test more.

Advantages
* Incremental Scala compile works
* Compiler-bridge fetching is done via coursier instead of the hacky jar
* Scala can be bumped beyond 2.12.8

Ultimately I'd like to go back to Bloop and use BSP with named pipes, but for now this solves several problems.

@mmjconolly @richardxia I'd appreciate reviews overall, but in particular I would like to see reviews of the second commit (message "Propagate errors in ScalaModule compile"). I am constrained that the API for `compileScalaModule` returns a `Path` so this still using `BadPath`, I'll change that some day with a package versioned API. I couldn't get `require` working so I did some manual `rmap` stuff, I'm curious how I can do it better.

jackkoenig committed:
Add support for incremental compile

* Switch from bloop to mill for underlying compilation mechanism
* Remove scala-bridge-fetcher

jackkoenig committed:
Propagate errors in ScalaModule compile

mmjconolly commented on scala.wake:
```diff
@@ -443,38 +443,43 @@ target memoizeMillCompile name \ module =
+        | editPlanEnvironment ("COURSIER_CACHE={relIvyCache}", visibleHashHack, _)
+        | setPlanDirectory buildDir
+        # Mill uses the entire classpath for caching so it must be local only for incremental compile
+        | setPlanLocalOnly True
```
Does this still need to be `LocalOnly`?

mmjconolly replied:
Oh I just noticed the comment above, can you expand a little on this, does it mean that the whole classpath should be in the visible files list?

jackkoenig replied:
Yeah, and it's good you asked because I forgot and had to remind myself which means it needs to be in the comment:

mill uses the filepaths in determining file identity, so `.fuse` shows up if it's not `LocalOnly`. With alternative sandboxing, it would not need to be `LocalOnly`.

mmjconolly replied:
`.fuse/pid` shouldn't be visible anymore with wake 0.19+ as we now mount the .fuse/pid directory over the top of the workspaces

See: https://github.com/sifive/wake/pull/444/files#diff-b61ba27252888d488d262f54042f9357R188

richardxia replied:
👍 on not making it LocalOnly.

jackkoenig commented:
I think there's a bug in how resources are handled, ran into it while using this on a complex project

richardxia commented on scala.wake:
```diff
@@ -337,143 +346,142 @@ global def getScalaModuleTransIvyDeps module =
+def millBuildDir module =
+  "{root.getPathName}/{module.getScalaModuleName}"
+
+target millBuildFile name \ module =
```
Do you really need to use the target subkey functionality here? I may just have PTSD from all the target subkey mismatch errors we hit, but when I look at this function, nothing here depends on the unserializable part of the ScalaModule (the `FnGeneratedSources` closure), so couldn't you memoize based on just the things that `genMillBuild` actually depends on?

You can always create a convenience function around it, but you don't need the function which accepts the `ScalaModule` itself to be memoizable. E.g.

```wake
target longOperation serializableArg1 serializableArg2 ... = ...
def myFunction unserializableArg =
  longOperation unserializableArg.getArg1 unserializableArg.getArg2
```

richardxia commented on scala.wake:
```diff
@@ -337,143 +346,142 @@ global def getScalaModuleTransIvyDeps module =
+      millBin, buildFile, depJars ++ sourceFiles ++ (ivyCacheDeps Unit)
+      | sortBy (_.getPathName <* _.getPathName) # Sorted to ensure hash is stable
+    depJarsResult
+    | rmap addOtherInputs
```
Many of these `rmap`s are hard to avoid because your actual return type for this function is `Path`, and the `require` construct is not easy to use for destructuring the error out of a Fail in the `else` case. For example, you can't do:

```wake
def maybeFoo = Fail "error".makeError
# This is pseudo-Scala code on the right half
require Pass foo = maybeFoo else { case Fail(e) => e.makeBadPath }
```

Even if you know that in the `else` case the `maybeFoo` is a `Fail`, you can't access the thing it has wrapped underneath.

I think what would probably simply things is splitting this into two functions, an outer lair that returns `Path` and an inner layer that returns `Result Path Error`. In the inner layer, you can always just return the `Fail` through the use of `require`, and in the outer layer, it's just one `match` expression.

Then you can do things like:

```wake
require Pass depJars = transDeps | map compileScalaModule | map getPathResult | findFail
def inputsResult = addOtherInputs depJars
def finalReturnValue =
  pretty
  | much
  | all
  | the
  | other stuff
Pass finalReturnValue
```

jackkoenig committed:
Update to Mill 0.7.4

Appears to fix issues where interrupting a compilation could cause
future compilations to fail

jackkoenig committed:
Fetch mill as part of fetchScala

jackkoenig committed:
Add dependency on scala-reflect 2.13.0

jackkoenig committed:
Add scala-reflect 2.13.0 source dependency fetching

jackkoenig committed:
Switch dependency from scala-reflect to scala-compiler

scala-compiler depends on scala-reflect

jackkoenig committed:
Try Zinc dep and compiler-bridge source dep

jackkoenig commented:
Closed in favor of https://github.com/sifive/api-scala-sifive/pull/53
